kernel/src/scan/mod.rs

```diff
@@ -264,15 +267,17 @@ impl Scan {
 
         let global_state = Arc::new(self.global_scan_state());
         let scan_data = self.scan_data(engine)?;
-        let mut scan_files = vec![];
-        for data in scan_data {
-            let (data, vec) = data?;
-            scan_files =
-                state::visit_scan_files(data.as_ref(), &vec, scan_files, scan_data_callback)?;
-        }
-        scan_files
-            .into_iter()
-            .map(|scan_file| -> DeltaResult<_> {
+        let scan_files_iter = scan_data
+            .map_ok(|(data, vec)| {
+                let scan_files = vec![];
+                state::visit_scan_files(data.as_ref(), &vec, scan_files, scan_data_callback)
+                    .map(IntoIterator::into_iter)
+            })
+            .flatten_ok()
+            .flatten_ok();
```

Comment:
We need one `flatten_ok()` to for the vec we created... but what is the other one collapsing?

Reply:
First we have scan_data, which is `Iterator<DeltaResult<_>>`. We use `map_ok` to map each `Ok` to `DeltaResult<Iterator<ScanFile>>`. So this becomes: `Iterator<DeltaResult<DeltaResult<Iterator<ScanFile>>>>`.

First flatten_ok:
`Iterator<DeltaResult<Iterator<ScanFile>>>`
Second flatten_ok:
`Iterator<DeltaResult<ScanFile>>`

Reply:
Darn you're right. Here's a minimal example: 
```
    let a = vec![Ok(1), Err("error"), Ok(4)].into_iter();
    let b = vec![Ok(5), Ok(6)].into_iter();
    let x = vec![Ok(a), Err("error"), Ok(b)].into_iter();
    for y in x.flatten_ok().flatten_ok() {
        println!("y: {:?}", y);
    }
```
Results in: 
```
y: Ok(1)
y: Ok(4)
y: Err("error")
y: Ok(5)
y: Ok(6)
```
One would expect to instead see two errors. I'll look into patching this. Thank you so much for flagging this @scovich!

Reply:
I've patched the PR to use `map_and_then` in scan. I updated a bunch of other places where I was using `map_ok` to instead do `map(|res| res.and_then(|value| {...})`. The reason I don't use `map_and_then` is that it is defined in the kernel crate, and is not accessible in test crates. I didn't make it public because it doesn't seem relevant to what  the kernel does, and it shouldn't be something users depend on.

Reply:
Also for those looking to sanity check, here's what `map_and_then` looks like in action:
```
    let a = vec![Ok(1), Err("error"), Ok(4)].into_iter();
    let b = vec![Ok(5), Ok(6)].into_iter();
    let x = vec![Ok(a), Err("error"), Ok(b)].into_iter();
    let iter = x
        .flatten_ok() // Iterator<Result<Iterator<Result<usize>>>> to Iterator<Result<Result<usize>>>
        .map_and_then(identity); // Iterator<Result<Result<usize>>> to Iterator<Result<usize>>
    for y in iter {
        println!("y: {:?}", y);
    }
```
Returns 
```
y: Ok(1)
y: Err("error")
y: Ok(4)
y: Err("error")
y: Ok(5)
y: Ok(6)
```